### PR TITLE
[#3] Basic configuration types and label rules

### PR DESCRIPTION
**Pull request by yaman**
Basic data types for a simple configuration.

### CODE REVIEW

**Comment by kahlil29** [comment]
Seems like the build is failing. Is it compiling fine on your local machine @yaman ?

**Comment by yaman** [comment]
@kahlil29 Yeap, it compiles fine.

**Review by kahlil29 — COMMENTED** [review]
- Not sure whether we should be using `String` or `Text`. Of late, I have become biased towards `Text` unless some manipulations are required which would be better via `String`s. 
@chshersh or @vrom911 would be able to confirm about this.

File: src/Crocodealer/Config.hs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+data LabelRule
+  = Create
+  | Override
+
+type Package = String
+
+data Config = Config
+  { username        :: Maybe String
+  , repository      :: Maybe String
+  , labelRules           :: [LabelRule]
```
Comment (kahlil29) [inline src/Crocodealer/Config.hs]: seems like a minor formatting issue here
```suggestion
  , labelRules      :: [LabelRule]
```

**Comment by yaman** [comment]
> * Not sure whether we should be using `String` or `Text`. Of late, I have become biased towards `Text` unless some manipulations are required which would be better via `String`s.
>   @chshersh or @vrom911 would be able to confirm about this.

Sure, I would be happy to change it.

**Comment by yaman** [comment]
Build fails because;
in `Crocodealer/Config.hs`, we have our types but since we do not use them **yet**, `stack build` with flag `--ghc-options=-Werror` brakes the build.

File: src/Crocodealer/Config.hs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  ) where
+
+data Label = Label
+  { name        :: String
+  , description :: String
+  , color       :: String -- Data.Colour.sRGB24read can convert string specifications like "#00aaff" or "00aaff" 
+  }
+
+data LabelRule
+  = Create
```
Comment (kahlil29) [inline src/Crocodealer/Config.hs]: I think that this should be of the form 

```
data LabelRule = Create Label | Override <something here>
``` 

In the case of `Override`, from the example that Dmitrii has provided in the issue description, it seems like a lambda needs to be provided in order to provide the logic of what is to be done to the existing label name.
But I'm not sure how that would fit in here.

File: src/Crocodealer/Config.hs (hunk 1 above)
Comment (yaman) [reply src/Crocodealer/Config.hs]: Seems I failed to run stylish-haskell continuously. Fixing it now..

File: src/Crocodealer/Config.hs (hunk 2 above)
Comment (yaman) [reply src/Crocodealer/Config.hs]: I might need some guidance on this one.

File: src/Crocodealer/Config.hs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module Crocodealer.Config
+  ( LabelRule
```
Comment (vrom911) [inline src/Crocodealer/Config.hs]: We are using this style guide over all of our projects: https://kowainik.github.io/posts/2019-02-06-style-guide
So by its rules, here we would have 7 spaces before the open bracket.

File: src/Crocodealer/Config.hs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  , Label
```
Comment (vrom911) [inline src/Crocodealer/Config.hs]: I guess you would need to import all of the constructors and fields for each data type here.

File: src/Crocodealer/Config.hs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  , Package
+  , Config
+  ) where
+
+data Label = Label
+  { name        :: String
```
Comment (vrom911) [inline src/Crocodealer/Config.hs]: We are using 4-spaces indentation rule by our style guide.

Comment (vrom911) [reply src/Crocodealer/Config.hs]: Also, @kahlil29 is right, it's better to use `Text` fields here and further :+1:

Comment (vrom911) [reply src/Crocodealer/Config.hs]: Fields names should be started with the data type name prefix:
```suggestion
  { labelName        :: !Text
```

**Review by vrom911 — COMMENTED** [review]
Nice start, @yaman ! I have some recommendations though

File: crocodealer.cabal [hunk 6]
```diff
@@ -23,6 +23,7 @@ source-repository head
 library
   hs-source-dirs:      src
   exposed-modules:     Crocodealer
+                       Crocodealer.Config
```
Comment (chshersh) [inline crocodealer.cabal]: It's not necessary required, but I prefer to indent hierarchical modules in the `.cabal` file as well for readability. See `issue-wanted` for example:

* https://github.com/kowainik/issue-wanted/blob/e612a87454b207fa47f64c630b17d014cdc4aa8a/issue-wanted.cabal#L61-L81

```suggestion
                           Crocodealer.Config
```

File: src/Crocodealer/Config.hs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  , color       :: String -- Data.Colour.sRGB24read can convert string specifications like "#00aaff" or "00aaff" 
+  }
+
+data LabelRule
+  = Create
+  | Override
+
+type Package = String
+
+data Config = Config
```
Comment (chshersh) [inline src/Crocodealer/Config.hs:13]: I would propose to separate these two concepts in two different modules:

1. `Crocodealer.Core.Label` for all label types and pure functions
2. `Crocodealer.Config` for the configuration data type and it's codec.

File: src/Crocodealer/Config.hs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  ( LabelRule
+  , Label
+  , Package
+  , Config
+  ) where
+
+data Label = Label
+  { name        :: String
+  , description :: String
+  , color       :: String -- Data.Colour.sRGB24read can convert string specifications like "#00aaff" or "00aaff"
```
Comment (chshersh) [inline src/Crocodealer/Config.hs]: We definitely should use more typesafe types here.

1. `newtype Name a = Name { unName :: Text }` and `Name Label` for the `labelName` field
2. `Text` for description is okay
3. `newtype HexColor = HexColor { unHexColor :: Text }` for `labelColor`. Constructor of `HexColor` shouldn't not exported. Instead, you need to create smart constructor:

```haskell
mkHexColor :: Text -> Maybe HexColor
```

Which verifies that the given text is in proper format: `#00ffaa` or something like that.

File: src/Crocodealer/Config.hs (hunk 2 above)
Comment (chshersh) [reply src/Crocodealer/Config.hs]: @kahlil29 is right, `LabelRule` should not be just enum. It should be:

```haskell
data LabelRule
    = Create Label
    | Override (Name Label) (Name Label)
```

First element of the `Override` constructor is `from` and the second is `to`. We change only names of the labels in the `Override` and then `Create` will actually do `create or update` work.

File: src/Crocodealer/Config.hs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  { name        :: String
+  , description :: String
+  , color       :: String -- Data.Colour.sRGB24read can convert string specifications like "#00aaff" or "00aaff" 
+  }
+
+data LabelRule
+  = Create
+  | Override
+
+type Package = String
```
Comment (chshersh) [inline src/Crocodealer/Config.hs]: This should be: 

```haskell
newtype Repo = Repo { unRepo :: Text }
```

**Review by chshersh — CHANGES_REQUESTED** [review]
The start is good, but some adjustments need to be made :slightly_smiling_face:

**Comment by vrom911** [comment]
Hey @yaman, do you need any help with the comments here?
Don't worry if you don't have time now, just let us know if we could help you :slightly_smiling_face:

**Comment by yaman** [comment]
I was just outside of the city last week. But I'm back and ready to fix according to feedbacks :blush:

File: src/Crocodealer/Config.hs (hunk 4 above)
Comment (yaman) [reply src/Crocodealer/Config.hs]: I am not sure if I undestood this one. Can you elaborate more please Veronika?

File: src/Crocodealer/Config.hs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+module Crocodealer.Config
+       ( Repo(..)
+       , Config(..)
```
Comment (chshersh) [inline src/Crocodealer/Config.hs]: I usually prefer to add and extra space in between data type and its exposed constructors
```suggestion
       , Config (..)
```

File: src/Crocodealer/Config.hs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+       ) where
+
+import Crocodealer.Core.Label
```
Comment (chshersh) [inline src/Crocodealer/Config.hs]: Let's add explicit import list here

File: src/Crocodealer/Config.hs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (chshersh) [inline src/Crocodealer/Config.hs:8]: There should be two empty lines after the import section according to our style guide
```suggestion


```

File: src/Crocodealer/Config.hs [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+newtype Repo = Repo
+    { unRepo :: Text
+    }
+
+data Config = Config
+    { username            :: Maybe Text
+    , repository          :: Maybe Text
```
Comment (chshersh) [inline src/Crocodealer/Config.hs]: All fields in this data type should be strict. Also, they should be prefixed with the data type name. Like `configUsername` and `configRepository`

File: src/Crocodealer/Core/Label.hs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+newtype HexColor = HexColor
+    { unHexColor :: Text
+    }
+
+data Label = Label
+    { labelName        :: Name Label
+    , labelDescription :: !Text
+    , labelColor       :: HexColor
```
Comment (chshersh) [inline src/Crocodealer/Core/Label.hs]: All fields should be strict, not only one :slightly_smiling_face:

File: src/Crocodealer/Core/Label.hs [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+data LabelRule
+    = Create Label
+    | Override (Name Label) (Name Label)
```
Comment (chshersh) [inline src/Crocodealer/Core/Label.hs]: Would be great to add some documentation to this data type. This is the only non-trivial data type in this PR, so it's always good to add some docs to it.

File: src/Crocodealer/Config.hs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+       ) where
+
+import Crocodealer.Core.Label
+
+newtype Repo = Repo
+    { unRepo :: Text
+    }
+
+data Config = Config
+    { username            :: Maybe Text
```
Comment (vrom911) [inline src/Crocodealer/Config.hs]: Could we using the prefix-base naming convention for every field here? Also, we can make all the fields strickt in this data type, like here:

```suggestion
    { configUsername            :: !(Maybe Text)
```

**Review by chshersh — CHANGES_REQUESTED** [review]
This looks nice! Code improved a lot and the structure is more clear now :slightly_smiling_face: I have some suggestions though, but we are moving forward :arrow_forward:

File: src/Crocodealer/Config.hs [hunk 17]
```diff
@@ -0,0 +1,17 @@
+module Crocodealer.Config
+       ( Repo(..)
```
Comment (vrom911) [inline src/Crocodealer/Config.hs]: There should be a space here :slightly_smiling_face: 
```suggestion
       ( Repo (..)
```

File: src/Crocodealer/Core/Label.hs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+newtype Name a = Name
+    { unName :: Text
+    }
+
+newtype HexColor = HexColor
+    { unHexColor :: Text
+    }
+
+data Label = Label
+    { labelName        :: Name Label
```
Comment (vrom911) [inline src/Crocodealer/Core/Label.hs]: You can make all of the fields strict

**Review by vrom911 — COMMENTED** [review]
Nice! I think derive `Show` at least would be helpful for future work and testing.

**Review by vrom911 — APPROVED** [review]
Thanks!

**Review by chshersh — APPROVED** [review]

**Pull request merged**